### DISCUSSION TITLE
What's the purpose of `clefs` array in the beginning of the measures array?

### DISCUSSION THREAD

**Discussion by GeorgeTailor** [w3c/mnx#343] [Q&A] [answered]
I'm having a hard time understanding the logic behind such structure.

Can clef sign change in the middle of the score? 
Can a single part have multiple clef signs? If yes, why then it is not reflected in the example of [organ](https://w3c.github.io/mnx/docs/mnx-reference/examples/organ-layout/)? This seems like a good example of such thing. But this leads to another question: how would mnx encode which measures belong to which clef and how to separate them? This doesn't seem to be possible at the moment.

If a clef sign can be only one at the beginning of the part and there can be only one clef sign for a part why it's not under `part` object on the same level as `measures` and is not an object instead of an array?

**Comment by adrianholovaty** [comment, marked as answer]
Hello — please see the example document called "Clef changes":
https://w3c.github.io/mnx/docs/mnx-reference/examples/clef-changes/

That `"clefs"` array is a list of clefs with their rhythmic positions — [positioned clef objects](https://w3c.github.io/mnx/docs/mnx-reference/objects/positioned-clef/).

Yes, it is indeed possible for the clef to change in the middle of a score, even in the middle of a measure. Our lives would be much easier if this weren't the case.

We haven't yet specified how to encode clefs for multi-stave instruments, such as a piano grand staff. MusicXML does this via the ["number" attribute on the clef object](https://w3c.github.io/musicxml/musicxml-reference/elements/clef/), and I imagine MNX will do something similar.

(I'm going to convert this to a GitHub discussion thread, since it's more of a question as opposed to a specific to-do.)

**Comment by GeorgeTailor** [comment]
I wish there was some kind of `mnx-lite` version, which doesn't support all of the possible corner-cases encountered in music sheets and instead focuses just on the most common ones. Are there any plans for such thing, maybe @adrianholovaty ?

**Reply by adrianholovaty** [reply]
No current plans for that, no. I think other formats would be a better fit for that, such as [ABC](https://abcnotation.com/).

**Reply by GeorgeTailor** [reply]
I'm afraid no major software programs support that at the moment, and given that ABC was around for quite some time, I doubt they will.
MusicXML, however, is quite wide spread, although most of the software produces mostly invalid musicXML due to complexity

**Reply by ahankinson** [reply]
It's not the format that's complex, it's music notation itself. You think there's a "simple" subset of music notation, but after 60 years of trying to invent such a format, nobody's really figured out what that subset is yet. The format is either too simple that nobody uses it because it isn't expressive enough outside of very specific use cases, or so expressive that the complexity of the format is a barrier to understanding it without a lot of study.

Maybe we should just ditch it all and go back to [DARMS](https://esf.ccarh.org/ccarh-wiki/DARMS-1976.pdf). 😄 

<img width="1496" alt="image" src="https://github.com/w3c/mnx/assets/163183/70d9383c-f60b-4c02-8bc3-db2d09564195">

**Reply by clnoel** [reply]
No, no, not DARMS! (Actually, I could probably cope with that. Musicnotes still relies on its history with DARMS in more ways than I would really like to admit. For instance, the shortcut for making a tie in our in-house music editor is 'J'.)

--Christina

**Reply by GeorgeTailor** [reply]
actually, there is such a format, and we are all familiar with it, it's a tablature. Tablature can cover all fretted instruments easily (fretless instruments like viola, violin etc. can also apply quite easily), there also exist variants for drums and brass instruments. There are also tablatures for key instruments like piano.

The only problem with tablature is a lack of explicit notes lengths, in this case a standard musical notation simplifies reading those quite a bit.

But having a tablature and standard musical notation which reflects only what is possible in a tablature is, imo, sophisticated enough to cover "the basics", where simplified standard musical notation and tablatures would complement each other.

**Reply by ahankinson** [reply]
That’s not an encoding format, that’s just another type of music notation. And it is just as complicated as “regular” notation once you get into all the different variants of lute tab (German, French, Italian, English) and organ tabs, etc. etc. And alternate tunings

If you build a machine readable  encoding format that handles tablature you won’t solve the complexity problem, you’ll simply shift it to a different domain.

**Reply by GeorgeTailor** [reply]
on the other hand to make a data informed decision we could scrape the internet for encoded musical notations in different formats and prepare a list of most commonly used "features" of musical notation, to satisfy over 90% of sheets out there and avoid some not that pretty structures just to satisfy a couple of percent of sheets out there.
Just to avoid having to deal with abominations like these:

![th-3267128972](https://github.com/w3c/mnx/assets/6958936/8a0bf15c-2391-40d0-afe1-29af7ec2dbc0)
![171422528-263191501](https://github.com/w3c/mnx/assets/6958936/6a3c4b81-4d51-4a95-a612-f169f4c733ed)


again, this is something for the future, if MNX turns out to be great success, then there will be no need in mnx-lite of any form

**Reply by GeorgeTailor** [reply]
my thinking here is that MusicXML as well as MNX (at least the final future version) will be able to encode the above sheets, but how many programs will be able to actually produce a valid and identical(!) MusicXML or MNX file?

**Reply by ahankinson** [reply]
I suspect you would quickly get caught up in the problems of averages and statistics. The most common, by far, is the ability to encode notes. That's not actually the problem. It's that individual scores sometimes require a very specific ability, and while those are the "long tail" of features, they're actually requirements for any given composer / repertoire / genre. 

I remember being at a conference where we discussed for at least a couple hours whether a slur should end ON the note, or AFTER a note, for a Chopin piece. The performers were adamant that it made a huge difference in how you perform the piece, so the ability to accurately position the ending of a slur was of vital importance to them.

Would that feature make the cut? Probably not. 99% of music notation doesn't need that. But if you're doing a Chopin edition (aka "bog-standard romantic piano music") it can be very important.

Similarly, I don't actually know of any "pure" *music* encoding format that would deal with those George Crumb (?) scores. You have to shift to a more graphical format for a dedicated piece of software, like MSCZ, Lilypond, or SCORE, to get the level of visual precision you need for those scores. MEI or MusicXML would both mess up the formatting, particularly if you tried to go from, say, MuseScore to Sibelius. It might be *theoretically* possible to do it in those formats, but you would have to study them for a very long time, and probably write your own dedicated exporter, to do it. 

If you want a simple notation format, I would look at **kern. 

https://verovio.humdrum.org

The advantage is that the learning curve isn't quite as steep as MusicXML or MEI. Very simple pieces are very simple to encode, but you can also do some pretty complex things as well.

**Reply by mscuthbert** [reply]
Bryan Ferneyhough. Second is Quartet -- 3?. First is Mnemosyne perhaps?

I don't see a need for a system to absolutely be able to represent everything here but I like the pieces so it would be good to allow them.

**Reply by GeorgeTailor** [reply]
I understand your point, but the question here is rather this:

does MNX strives to be a guideline on how to encode musical sheets or a music exchange format? The latter suggests that you can export it in one piece of software and import it in another one and it should just work. But from my experience, most software programs were not able to fully read the musicXML file they produced, the imported exported file was different than the native format for any particular program. Not even mentioning trying to import exported file from/to another piece of software.

I appreciate you suggesting other formats but they are not exchangeable, because no one is using them apart from the owner. Btw, I tried to export https://verovio.humdrum.org/ as a musicXML file and import it into GuitarPro, the beams were not properly encoded, the instrument played in GuitarPro was 4 guitars instead of piano, also the tempo is different, 100 bpm in guitarPro and a little more in this humdrum thingy.

IMO, you can choose only 1: either support everything and it's not exchangeable, or support a subset with high chances it will be exchangeable due to simplicity and ease of enforcement of rules via some kind of schema.

**Reply by adrianholovaty** [reply]
When it comes to music exchange, there are many possible points of failure. If exporting from Software A to Software B in MusicXML, any or all of the following problems might happen:

1. **Software A doesn't export a particular aspect of notation.** (Example: it doesn't bother to encode the graphical position of each glyph.)
2. **Software A _does_ export a particular aspect of notation but its export implementation is buggy.** (Maybe it's a simple oversight in programming, or maybe it's due to confusion over the complicated MusicXML spec.)
3. **Software A _does_ export a particular aspect of notation but its internal model of that notation doesn't match the model of MusicXML.** (Example: when exporting chord names, MusicXML requires the data to be split into subelements "step", "kind", "degree", etc., whereas Software A might just store the whole thing as a string and cannot reliably split that data.)
4. **MusicXML doesn't provide an unambiguous way to encode a particular aspect of notation** (i.e., it's interpreted differently by different applications).
5. **Software B doesn't import a particular aspect of notation.** (Example: many notation programs don't bother to import positional information, as they decide to make these calculations independently.)
6. **Software B doesn't even _support_ a particular aspect of notation.** (Examples: nested tuplets, beams over barlines, inner-bar clef changes.) In some cases this is OK, in other cases it leads to outright incorrect or broken notation upon import.
7. **Software B _does_ import a particular aspect of notation, but its import implementation is buggy.** (Maybe it's a simple oversight in programming, or maybe it's due to confusion over the complicated MusicXML spec.)
8. **Software B _does_ import a particular aspect of notation, but its internal model of that notation doesn't match the model of MusicXML.**

In my view, MNX is primarily concerned with problem 4. We also think long and hard about how to avoid problems 2, 3, 7 and 8. Of course, we think about _all_ of these problems, but some of them are inherently out of our control. The best we can do is to design a format that's unambiguous whenever possible, with documentation that's as clear as possible.

**Reply by ahankinson** [reply]
> IMO, you can choose only 1: either support everything and it's not exchangeable, or support a subset with high chances it will be exchangeable due to simplicity and ease of enforcement of rules via some kind of schema.

You are exactly right. The central issue, though, is **whose** simplicity are we going to agree on? Your simple requirements, and my simple requirements, and Myke's simple requirements, are not the same. 

We're going through this *exact* conversation over on the MEI github where we're trying to work out what constitutes "Basic" encoding. Is the ability to arbitrarily scale a staff a "basic" feature? [Some would say so](https://github.com/music-encoding/music-encoding/issues/1444). What about the ability to give the staff lines a specific colour? Or to mark the staff lines visible or not? [Or the ability](https://github.com/music-encoding/music-encoding/issues/1324) to shift a rest up and down the staff on a given line? 

It's not like we (the designers of notation encoding formats) set out to purposefully create complex systems where enforcement of the rules is difficult. In fact, I would hazard a guess that the motive behind most new formats is to try and address this exact problem. 

My take on it is that, fundamentally, music notation is a picture that represents some system of sound making. We *think* music notation is like text, but it really isn't -- it's what you get when you try and draw sound. So a notation **format** is a bit like trying to invent a format that will accurately represent a visual phenomenon,  and then we impose the requirements that it must represent everything from a Dutch Master to an Impressionist to a Jackson Pollack to a couple scribbles on the back of a napkin, and do that *semantically* (that is, encode the semantics of the contents of the painting, rather than simply the shapes and colours _a la_ SVG). There is no way that can be done and keep precise fidelity with all. And not impose any mutually exclusive behaviours. 

**kern is exchangeable. [music21](https://web.mit.edu/music21/doc/moduleReference/moduleHumdrumSpineParser.html) supports it. [Verovio](https://book.verovio.org/toolkit-reference/input-formats.html#humdrum) supports it. It's an "open" format, insofar that there's no proprietary motive behind it. It may not be exchangeable in the tools you want to use, but there are [very](https://josquin.stanford.edu) [significant](https://polishscores.org) [bodies](http://kern.ccarh.org) of encoded notation available in it. Sibelius, Finale, and MuseScore might be the household names, but there's really a lot of stuff happening outside of those tools.
